File: smartcontract/programs/doublezero-serviceability/src/pda.rs
use std::net::Ipv4Addr;

use solana_program::pubkey::Pubkey;

use crate::{
    seeds::{
        SEED_ACCESS_PASS, SEED_CONFIG, SEED_CONTRIBUTOR, SEED_DEVICE, SEED_DEVICE_TUNNEL_BLOCK,
        SEED_DZ_PREFIX_BLOCK, SEED_EXCHANGE, SEED_GLOBALSTATE, SEED_LINK, SEED_LINK_IDS,
        SEED_LOCATION, SEED_MULTICASTGROUP_BLOCK, SEED_MULTICAST_GROUP, SEED_PREFIX,
        SEED_PROGRAM_CONFIG, SEED_SEGMENT_ROUTING_IDS, SEED_TENANT, SEED_TUNNEL_IDS, SEED_USER,
        SEED_USER_TUNNEL_BLOCK, SEED_VRF_IDS,
    },
    state::user::UserType,
};

pub fn get_globalstate_pda(program_id: &Pubkey) -> (Pubkey, u8) {
    Pubkey::find_program_address(&[SEED_PREFIX, SEED_GLOBALSTATE], program_id)
}

pub fn get_globalconfig_pda(program_id: &Pubkey) -> (Pubkey, u8) {
    Pubkey::find_program_address(&[SEED_PREFIX, SEED_CONFIG], program_id)
}

pub fn get_program_config_pda(program_id: &Pubkey) -> (Pubkey, u8) {
    Pubkey::find_program_address(&[SEED_PREFIX, SEED_PROGRAM_CONFIG], program_id)
}

pub fn get_location_pda(program_id: &Pubkey, index: u128) -> (Pubkey, u8) {
    Pubkey::find_program_address(
        &[SEED_PREFIX, SEED_LOCATION, &index.to_le_bytes()],
        program_id,
    )
}

pub fn get_exchange_pda(program_id: &Pubkey, index: u128) -> (Pubkey, u8) {
    Pubkey::find_program_address(
        &[SEED_PREFIX, SEED_EXCHANGE, &index.to_le_bytes()],
        program_id,
    )
}

pub fn get_device_pda(program_id: &Pubkey, index: u128) -> (Pubkey, u8) {
    Pubkey::find_program_address(
        &[SEED_PREFIX, SEED_DEVICE, &index.to_le_bytes()],
        program_id,
    )
}

pub fn get_link_pda(program_id: &Pubkey, index: u128) -> (Pubkey, u8) {
    Pubkey::find_program_address(&[SEED_PREFIX, SEED_LINK, &index.to_le_bytes()], program_id)
}

pub fn get_user_old_pda(program_id: &Pubkey, index: u128) -> (Pubkey, u8) {
    Pubkey::find_program_address(&[SEED_PREFIX, SEED_USER, &index.to_le_bytes()], program_id)
}

pub fn get_user_pda(program_id: &Pubkey, ip: &Ipv4Addr, user_type: UserType) -> (Pubkey, u8) {
    Pubkey::find_program_address(
        &[SEED_PREFIX, SEED_USER, &ip.octets(), &[user_type as u8]],
        program_id,
    )
}

pub fn get_multicastgroup_pda(program_id: &Pubkey, index: u128) -> (Pubkey, u8) {
    Pubkey::find_program_address(
        &[SEED_PREFIX, SEED_MULTICAST_GROUP, &index.to_le_bytes()],
        program_id,
    )
}

pub fn get_contributor_pda(program_id: &Pubkey, index: u128) -> (Pubkey, u8) {
    Pubkey::find_program_address(
        &[SEED_PREFIX, SEED_CONTRIBUTOR, &index.to_le_bytes()],
        program_id,
    )
}

pub fn get_tenant_pda(program_id: &Pubkey, code: &str) -> (Pubkey, u8) {
    Pubkey::find_program_address(&[SEED_PREFIX, SEED_TENANT, code.as_bytes()], program_id)
}

pub fn get_accesspass_pda(
    program_id: &Pubkey,
    client_ip: &Ipv4Addr,
    user_payer: &Pubkey,
) -> (Pubkey, u8) {
    Pubkey::find_program_address(
        &[
            SEED_PREFIX,
            SEED_ACCESS_PASS,
            &client_ip.octets(),
            &user_payer.to_bytes(),
        ],
        program_id,
    )
}

pub fn get_resource_extension_pda(
    program_id: &Pubkey,
    resource_type: crate::resource::ResourceType,
) -> (Pubkey, u8, &'static [u8]) {
    match resource_type {
        crate::resource::ResourceType::DeviceTunnelBlock => {
            let (pda, bump_seed) =
                Pubkey::find_program_address(&[SEED_PREFIX, SEED_DEVICE_TUNNEL_BLOCK], program_id);
            (pda, bump_seed, SEED_DEVICE_TUNNEL_BLOCK)
        }
        crate::resource::ResourceType::UserTunnelBlock => {
            let (pda, bump_seed) =
                Pubkey::find_program_address(&[SEED_PREFIX, SEED_USER_TUNNEL_BLOCK], program_id);
            (pda, bump_seed, SEED_USER_TUNNEL_BLOCK)
        }
        crate::resource::ResourceType::MulticastGroupBlock => {
            let (pda, bump_seed) =
                Pubkey::find_program_address(&[SEED_PREFIX, SEED_MULTICASTGROUP_BLOCK], program_id);
            (pda, bump_seed, SEED_MULTICASTGROUP_BLOCK)
        }
        crate::resource::ResourceType::DzPrefixBlock(ref associated_pk, index) => {
            let (pda, bump_seed) = Pubkey::find_program_address(
                &[
                    SEED_PREFIX,
                    SEED_DZ_PREFIX_BLOCK,
                    &associated_pk.to_bytes(),
                    &index.to_le_bytes(),
                ],
                program_id,
            );
            (pda, bump_seed, SEED_DZ_PREFIX_BLOCK)
        }
        crate::resource::ResourceType::TunnelIds(ref associated_pk, index) => {
            let (pda, bump_seed) = Pubkey::find_program_address(
                &[
                    SEED_PREFIX,
                    SEED_TUNNEL_IDS,
                    &associated_pk.to_bytes(),
                    &index.to_le_bytes(),
                ],
                program_id,
            );
            (pda, bump_seed, SEED_TUNNEL_IDS)
        }
        crate::resource::ResourceType::LinkIds => {
            let (pda, bump_seed) =
                Pubkey::find_program_address(&[SEED_PREFIX, SEED_LINK_IDS], program_id);
            (pda, bump_seed, SEED_LINK_IDS)
        }
        crate::resource::ResourceType::SegmentRoutingIds => {
            let (pda, bump_seed) =
                Pubkey::find_program_address(&[SEED_PREFIX, SEED_SEGMENT_ROUTING_IDS], program_id);
            (pda, bump_seed, SEED_SEGMENT_ROUTING_IDS)
        }
        crate::resource::ResourceType::VrfIds => {
            let (pda, bump_seed) =
                Pubkey::find_program_address(&[SEED_PREFIX, SEED_VRF_IDS], program_id);
            (pda, bump_seed, SEED_VRF_IDS)
        }
    }
}

File: smartcontract/sdk/go/serviceability/pda.go
package serviceability

import (
	"encoding/binary"

	"github.com/gagliardetto/solana-go"
)

// PDA seeds matching Rust implementation in seeds.rs
const (
	SeedPrefix                  = "doublezero"
	SeedGlobalState             = "globalstate"
	SeedGlobalConfig            = "config"
	SeedProgramConfig           = "programconfig"
	SeedLinkIds                 = "linkids"
	SeedSegmentRoutingIds       = "segmentroutingids"
	SeedUserTunnelBlock         = "usertunnelblock"
	SeedDeviceTunnelBlock       = "devicetunnelblock"
	SeedMulticastGroupBlock     = "multicastgroupblock"
	SeedMulticastPublisherBlock = "multicastpublisherblock"
	SeedTenant                  = "tenant"
	SeedPermission              = "permission"
	SeedUser                    = "user"
	SeedAccessPass              = "accesspass"
	SeedTunnelIds               = "tunnelids"
	SeedDzPrefixBlock           = "dzprefixblock"
)

// DeriveGlobalStatePDA derives the PDA for the GlobalState account.
func DeriveGlobalStatePDA(programID solana.PublicKey) (solana.PublicKey, uint8, error) {
	seeds := [][]byte{
		[]byte(SeedPrefix),
		[]byte(SeedGlobalState),
	}
	return solana.FindProgramAddress(seeds, programID)
}

// DeriveGlobalConfigPDA derives the PDA for the GlobalConfig account.
func DeriveGlobalConfigPDA(programID solana.PublicKey) (solana.PublicKey, uint8, error) {
	seeds := [][]byte{
		[]byte(SeedPrefix),
		[]byte(SeedGlobalConfig),
	}
	return solana.FindProgramAddress(seeds, programID)
}

// DeriveProgramConfigPDA derives the PDA for the ProgramConfig account.
func DeriveProgramConfigPDA(programID solana.PublicKey) (solana.PublicKey, uint8, error) {
	seeds := [][]byte{
		[]byte(SeedPrefix),
		[]byte(SeedProgramConfig),
	}
	return solana.FindProgramAddress(seeds, programID)
}

// GetGlobalStatePDA is a convenience alias for DeriveGlobalStatePDA.
func GetGlobalStatePDA(programID solana.PublicKey) (solana.PublicKey, uint8, error) {
	return DeriveGlobalStatePDA(programID)
}

// GetLinkIdsPDA derives the PDA for the global LinkIds resource extension
func GetLinkIdsPDA(programID solana.PublicKey) (solana.PublicKey, uint8, error) {
	seeds := [][]byte{
		[]byte(SeedPrefix),
		[]byte(SeedLinkIds),
	}
	return solana.FindProgramAddress(seeds, programID)
}

// GetSegmentRoutingIdsPDA derives the PDA for the global SegmentRoutingIds resource extension
func GetSegmentRoutingIdsPDA(programID solana.PublicKey) (solana.PublicKey, uint8, error) {
	seeds := [][]byte{
		[]byte(SeedPrefix),
		[]byte(SeedSegmentRoutingIds),
	}
	return solana.FindProgramAddress(seeds, programID)
}

// GetUserTunnelBlockPDA derives the PDA for the global UserTunnelBlock resource extension
func GetUserTunnelBlockPDA(programID solana.PublicKey) (solana.PublicKey, uint8, error) {
	seeds := [][]byte{
		[]byte(SeedPrefix),
		[]byte(SeedUserTunnelBlock),
	}
	return solana.FindProgramAddress(seeds, programID)
}

// GetDeviceTunnelBlockPDA derives the PDA for the global DeviceTunnelBlock resource extension
func GetDeviceTunnelBlockPDA(programID solana.PublicKey) (solana.PublicKey, uint8, error) {
	seeds := [][]byte{
		[]byte(SeedPrefix),
		[]byte(SeedDeviceTunnelBlock),
	}
	return solana.FindProgramAddress(seeds, programID)
}

// GetMulticastGroupBlockPDA derives the PDA for the global MulticastGroupBlock resource extension
func GetMulticastGroupBlockPDA(programID solana.PublicKey) (solana.PublicKey, uint8, error) {
	seeds := [][]byte{
		[]byte(SeedPrefix),
		[]byte(SeedMulticastGroupBlock),
	}
	return solana.FindProgramAddress(seeds, programID)
}

// GetMulticastPublisherBlockPDA derives the PDA for the global MulticastPublisherBlock resource extension
func GetMulticastPublisherBlockPDA(programID solana.PublicKey) (solana.PublicKey, uint8, error) {
	seeds := [][]byte{
		[]byte(SeedPrefix),
		[]byte(SeedMulticastPublisherBlock),
	}
	return solana.FindProgramAddress(seeds, programID)
}

// GetTenantPDA derives the PDA for a tenant account based on its code
func GetTenantPDA(programID solana.PublicKey, code string) (solana.PublicKey, uint8, error) {
	seeds := [][]byte{
		[]byte(SeedPrefix),
		[]byte(SeedTenant),
		[]byte(code),
	}
	return solana.FindProgramAddress(seeds, programID)
}

// GetPermissionPDA derives the PDA for a Permission account given the user_payer pubkey.
func GetPermissionPDA(programID solana.PublicKey, userPayer solana.PublicKey) (solana.PublicKey, uint8, error) {
	seeds := [][]byte{
		[]byte(SeedPrefix),
		[]byte(SeedPermission),
		userPayer[:],
	}
	return solana.FindProgramAddress(seeds, programID)
}

// GetUserPDA derives the PDA for a User account, keyed by (client_ip, user_type).
// Mirrors smartcontract/programs/doublezero-serviceability/src/pda.rs:get_user_pda.
func GetUserPDA(programID solana.PublicKey, clientIP [4]byte, userType UserUserType) (solana.PublicKey, uint8, error) {
	seeds := [][]byte{
		[]byte(SeedPrefix),
		[]byte(SeedUser),
		clientIP[:],
		{byte(userType)},
	}
	return solana.FindProgramAddress(seeds, programID)
}

// GetAccessPassPDA derives the PDA for an AccessPass account, keyed by (client_ip, user_payer).
// Mirrors smartcontract/programs/doublezero-serviceability/src/pda.rs:get_accesspass_pda.
func GetAccessPassPDA(programID solana.PublicKey, clientIP [4]byte, userPayer solana.PublicKey) (solana.PublicKey, uint8, error) {
	seeds := [][]byte{
		[]byte(SeedPrefix),
		[]byte(SeedAccessPass),
		clientIP[:],
		userPayer[:],
	}
	return solana.FindProgramAddress(seeds, programID)
}

// GetTunnelIdsPDA derives the PDA for a per-device TunnelIds resource extension at the given index.
// Rust uses usize (8 bytes on 64-bit) little-endian for the index; we always encode 8 bytes.
func GetTunnelIdsPDA(programID solana.PublicKey, devicePK solana.PublicKey, index uint64) (solana.PublicKey, uint8, error) {
	var idxBuf [8]byte
	binary.LittleEndian.PutUint64(idxBuf[:], index)
	seeds := [][]byte{
		[]byte(SeedPrefix),
		[]byte(SeedTunnelIds),
		devicePK[:],
		idxBuf[:],
	}
	return solana.FindProgramAddress(seeds, programID)
}

// GetDzPrefixBlockPDA derives the PDA for a per-device DzPrefixBlock resource extension at the given index.
// Rust uses usize (8 bytes on 64-bit) little-endian for the index; we always encode 8 bytes.
func GetDzPrefixBlockPDA(programID solana.PublicKey, devicePK solana.PublicKey, index uint64) (solana.PublicKey, uint8, error) {
	var idxBuf [8]byte
	binary.LittleEndian.PutUint64(idxBuf[:], index)
	seeds := [][]byte{
		[]byte(SeedPrefix),
		[]byte(SeedDzPrefixBlock),
		devicePK[:],
		idxBuf[:],
	}
	return solana.FindProgramAddress(seeds, programID)
}
